### ISSUE TITLE
Adding Processor TID to logging output.

### ISSUE THREAD

**Issue by jnarowski** [mperham/sidekiq#4810]
Adding process TID into logging.

**Currently, if you do a** 
kill -TTIN [PID] 

**You get an output like this:**
```
2021-02-11T15:41:49.998645+00:00 app[sidekiq.1]: {"ts":"2021-02-11T15:41:49.998Z","pid":95,"tid":"ou40t9a9r","lvl":"WARN","msg":"Thread TID-ou3wah1qj processor"}
2021-02-11T15:41:49.999125+00:00 app[sidekiq.1]: {"ts":"2021-02-11T15:41:49.998Z","pid":95,"tid":"ou40t9a9r","lvl":"WARN","msg":"/app/vendor/bundle/ruby/2.6.0
```
.....
In order to properly associate the output with a stuck job in Sidekiq, you need the "Thread TID-ou3wah1qj processor" part. The problem is, I don't have debug info from that line if I filter by **Thread TID-ou3wah1qj processor**. I propose we add ou3wah1qj to the logging output that includes the "msg", so I can easily filter on it using logging tools.

**Comment by mperham** [comment]
There is a tid element in your sample output. I’m confused?

**Comment by mperham** [comment]
Oh I see. That tid is for the Thread that is logging...

**Comment by jnarowski** [comment]
Yeah, I need **ou3wah1qj** in the second output to be able to quickly and reliably filter/process in my logging tool.

Something like: "ttid"

**EG**
```
2021-02-11T15:41:49.999125+00:00 app[sidekiq.1]: {"ts":"2021-02-11T15:41:49.998Z","pid":95,"tid":"ou40t9a9r","ttid":"ou3wah1qj","lvl":"WARN","msg":"/app
```

**Comment by mperham** [comment]
One issue is that you've chosen to use the JSON output format. Turn that off and it might be easier to read?

**Comment by jnarowski** [comment]
That's not the issue. The issue is the TID-ou3wah1qj is not in that line at all. Logging systems parse the entire line JSON or not and the fact that it's in the line above means it's totally disconnected from my ability to filter down and find information. It needs to be in the same block as the other info JSON or not. Does that make sense?

**Comment by manojnaidu619** [comment]
I'm happy to work on this👍

**Comment by liijunwei** [comment]
Sorry about bringing this old issue up again, just wondering whether this issue needs further discusstion.

I have a draft solution like following code snippet, and I'm glad to work on this.

here are several pending questions:

1. Is this solution good enough?
2. Shoud I make it configurable?
3. I don't really like this pattern matching solution, I guess better redesign `TTIN` signal handler or logging component(very vague idea, not sure if it's worth it)
    + potential issue: unstable dependency(the log message)
    + potential issue: might have performance issue using regexp matching for every log message
4. Only for JSON formatter?

```ruby
# lib/sidekiq/logger.rb
module Sidekiq
  class Logger < ::Logger
    module Formatters

      class JSON < Base
        def call(severity, time, program_name, message)
          hash = {
            ts: time.utc.iso8601(3),
            pid: ::Process.pid,
            tid: tid,
            lvl: severity,
            msg: message
          }
          c = ctx
          hash["ctx"] = c unless c.empty?

          ttid = /^Thread TID-(.*)\s{1}/.match(message)&.captures&.first
          hash["ttid"] = ttid unless c.empty?

          Sidekiq.dump_json(hash) << "\n"
        end
      end

    end
  end
end
```

**Comment by mperham** [comment]
Unfortunately fixing this would require a pretty major rework of the logging code. That's why it's been open for years.

**Comment by liijunwei** [comment]
Got it, thanks for replying, I think I'll put it on hold

**Comment by sobrinho** [comment]
@mperham this is stale and can be closed.
